pkg/pgmodel/ingestor/dispatcher.go

```diff
@@ -178,6 +189,250 @@ func (p *pgxDispatcher) Close() {
+	return id, err
+}
+
+func (p *pgxDispatcher) insertTags(ctx context.Context, tags map[string]interface{}, typ TagType) error {
+	for k, v := range tags {
+		_, err := p.conn.Exec(ctx, "SELECT "+schema.TracePublic+".put_tag_key($1, $2::"+schema.TracePublic+".tag_type)", k, typ)
+
+		if err != nil {
+			return err
+		}
+
+		val, err := json.Marshal(v)
+		if err != nil {
+			return err
+		}
+
+		_, err = p.conn.Exec(ctx, "SELECT "+schema.TracePublic+".put_tag($1, $2, $3::"+schema.TracePublic+".tag_type)",
+			k,
+			string(val),
+			typ,
+		)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getTraceStateValue(ts pdata.TraceState) (result pgtype.Text) {
+	if string(ts) == "" {
+		result.Status = pgtype.Null
+	} else {
+		result.String = string(ts)
+		result.Status = pgtype.Present
+	}
+
+	return result
+}
+
+func (p *pgxDispatcher) InsertSpanLinks(ctx context.Context, links pdata.SpanLinkSlice, traceID [16]byte, spanID [8]byte, spanStartTime time.Time) error {
+	spanIDInt := convertByteArrayToInt64(spanID)
+	for i := 0; i < links.Len(); i++ {
+		link := links.At(i)
+		linkedSpanIDInt := convertByteArrayToInt64(link.SpanID().Bytes())
+
+		rawTags := link.Attributes().AsRaw()
+		if err := p.insertTags(ctx, rawTags, LinkTagType); err != nil {
+			return err
+		}
+		jsonTags, err := json.Marshal(rawTags)
+		if err != nil {
+			return err
+		}
+		_, err = p.conn.Exec(ctx, "INSERT INTO "+schema.Trace+".link (trace_id, span_id, span_start_time, linked_trace_id, linked_span_id, trace_state, tags, dropped_tags_count, link_nbr) VALUES ( $1, $2, $3, $4, $5, $6, "+schema.TracePublic+".get_tag_map($7), $8, $9 )",
+			getUUID(traceID),
+			spanIDInt,
+			spanStartTime,
+			getUUID(link.TraceID().Bytes()),
+			linkedSpanIDInt,
+			getTraceStateValue(link.TraceState()),
+			string(jsonTags),
+			link.DroppedAttributesCount(),
+			i,
+		)
+
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *pgxDispatcher) InsertSpanEvents(ctx context.Context, events pdata.SpanEventSlice, traceID [16]byte, spanID [8]byte) error {
+	spanIDInt := convertByteArrayToInt64(spanID)
+	for i := 0; i < events.Len(); i++ {
+		event := events.At(i)
+		rawTags := event.Attributes().AsRaw()
+		if err := p.insertTags(ctx, rawTags, EventTagType); err != nil {
+			return err
+		}
+		jsonTags, err := json.Marshal(rawTags)
+		if err != nil {
+			return err
+		}
+		_, err = p.conn.Exec(ctx, "INSERT INTO "+schema.Trace+".event (time, trace_id, span_id, name, event_nbr, tags, dropped_tags_count) VALUES ( $1, $2, $3, $4, $5, "+schema.TracePublic+".get_tag_map($6), $7 )",
+			event.Timestamp().AsTime(),
+			getUUID(traceID),
+			spanIDInt,
+			event.Name(),
+			i,
+			string(jsonTags),
+			event.DroppedAttributesCount(),
+		)
+
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func convertByteArrayToInt64(buf [8]byte) int64 {
+	ux := binary.BigEndian.Uint64(buf[:])
+
+	x := int64(ux >> 1)
```

Comment:
I don't get this conversion code why can't we simply cast to an int64? Can you add a comment with explanation/link? Also we probably need the reverse operation.

Reply:
I'll add a comment and add a reverse function, export them both because one will be unused/dead code right now.

Reply:
Also, if we just cast to int64, it can overflow from an uint64.

Reply:
OK, my bad, after checking the conversion... seems like Go handles it nicely. I'm gonna remove this and just do a cast.